File: src/AppStudio/Design/DesignInitializer.cs
using PixUI;
using AppBoxCore;

namespace AppBoxDesign;

/// <summary>
/// 初始化设计时，eg:注册序列化器等
/// </summary>
[TSType("AppBoxDesign.DesignInitializer")]
public static class DesignInitializer
{
    static DesignInitializer()
    {
        // TypeSerializer.RegisterKnownType(new BinSerializer(PayloadType.EntityModelVO,
        //     typeof(EntityModelVO), () => new EntityModelVO()));
        // TypeSerializer.RegisterKnownType(new BinSerializer(PayloadType.EntityMemberVO,
        //     typeof(EntityMemberVO)));
        // TypeSerializer.RegisterKnownType(new BinSerializer(PayloadType.EntityFieldVO,
        //     typeof(EntityFieldVO), () => new EntityFieldVO()));
        // TypeSerializer.RegisterKnownType(new BinSerializer(PayloadType.EntityRefVO,
        //     typeof(EntityRefVO), () => new EntityRefVO()));
        // TypeSerializer.RegisterKnownType(new BinSerializer(PayloadType.EntitySetVO,
        //     typeof(EntitySetVO), () => new EntitySetVO()));
        // TypeSerializer.RegisterKnownType(new BinSerializer(PayloadType.EntityMemberInfo,
        //     typeof(EntityMemberInfo), () => new EntityMemberInfo()));
        // // TypeSerializer.RegisterKnownType(new BinSerializer(PayloadType.NewNodeResult,
        // //     typeof(NewNodeResult_OLD), () => new NewNodeResult_OLD()));
        // TypeSerializer.RegisterKnownType(new BinSerializer(PayloadType.CodeProblem,
        //     typeof(CodeProblem), () => new CodeProblem()));
        // TypeSerializer.RegisterKnownType(new BinSerializer(PayloadType.ChangedModel,
        //     typeof(ChangedModel), () => new ChangedModel()));
        // TypeSerializer.RegisterKnownType(new BinSerializer(PayloadType.OrderedField,
        //     typeof(OrderedField), () => new OrderedField()));
        // TypeSerializer.RegisterKnownType(new BinSerializer(PayloadType.ReferenceVO,
        //     typeof(ReferenceVO), () => new ReferenceVO()));
        // TypeSerializer.RegisterKnownType(new BinSerializer(PayloadType.TextChange,
        //     typeof(TextChange), () => new TextChange()));
    }

    public static Task TryInit() => Task.CompletedTask; //do nothing for native
}

File: src/Dev.Web/src/AppBoxDesign/HandWrite/DesignInitializer.ts
import {
    DesignTreeVO, CompletionItem, NewNodeResult, ChangedModel, CodeProblem,
    EntityModelVO, EntityFieldVO, EntityRefVO, EntitySetVO, EntityMemberInfo, ReferenceVO
} from '@/AppBoxDesign'
import {TSCSharpLanguage} from "@/CodeEditor";
import {FieldWithOrder, PayloadType, TypeSerializer} from "@/AppBoxCore";

export class DesignInitializer {
    public static async TryInit(): Promise<void> {
        if (TypeSerializer.GetSerializer(PayloadType.DesignTree) != null)
            return;

        // 初始化TreeSitter
        const TreeSitter: any = (<any>window).TreeSitter;
        await TreeSitter.init();
        let csharpLanguage = await TreeSitter.Language.load('/tree-sitter-c_sharp.wasm');
        TSCSharpLanguage.Init(csharpLanguage);

        // 注册序列化器(仅反序列化后端设计时类型)
        TypeSerializer.RegisterKnownType(PayloadType.EntityModelVO, false, () => new EntityModelVO());
        //TypeSerializer.RegisterKnownType(PayloadType.EntityMemberVO, false);
        TypeSerializer.RegisterKnownType(PayloadType.EntityFieldVO, false, () => new EntityFieldVO());
        TypeSerializer.RegisterKnownType(PayloadType.EntityRefVO, false, () => new EntityRefVO());
        TypeSerializer.RegisterKnownType(PayloadType.EntitySetVO, false, () => new EntitySetVO());
        TypeSerializer.RegisterKnownType(PayloadType.EntityMemberInfo, false, () => new EntityMemberInfo());
        TypeSerializer.RegisterKnownType(PayloadType.DesignTree, false, () => new DesignTreeVO());
        TypeSerializer.RegisterKnownType(PayloadType.NewNodeResult, false, () => new NewNodeResult());
        TypeSerializer.RegisterKnownType(PayloadType.CodeProblem, true, () => new CodeProblem());
        TypeSerializer.RegisterKnownType(PayloadType.CompletionItem, true, () => new CompletionItem());
        TypeSerializer.RegisterKnownType(PayloadType.ChangedModel, true, () => new ChangedModel());
        TypeSerializer.RegisterKnownType(PayloadType.FieldWithOrder, true, () => new FieldWithOrder());
        TypeSerializer.RegisterKnownType(PayloadType.ReferenceVO, false, () => new ReferenceVO());
    }
}